### PR TITLE
Improve TowerResolver performance, especially for REPL

### PR DESCRIPTION
**Pull request by ileasile**
Implemented definitelyDoesNotContainName in some scope implementations and added special annotation for skipping extensions resolving for implicits. Compiling times are now increasing lineary that allows to compile hundreds and thousands of snippets without serious performance degradation.

### CODE REVIEW

File: compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/TowerResolver.kt [hunk 1]
```diff
@@ -227,9 +229,16 @@ class TowerResolver {
             }
 
             // extension for implicit receiver
-            for (nonLocalLevel in nonLocalLevels) {
-                TowerData.BothTowerLevelAndImplicitReceiver(nonLocalLevel, implicitReceiver).process()?.let { return it }
-            }
+            val receiverValue = implicitReceiver.receiverValue
+            val resolveExtensions = receiverValue !is ImplicitClassReceiver ||
```
Comment (ligee) [inline compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/TowerResolver.kt:233]: The intentions will look much cleaner if you'll reverse the flag meaning, i.e. name it `skipResolveExtensions` and rewrite the code accordingly.

File: libraries/scripting/common/src/kotlin/script/experimental/api/scriptCompilation.kt [hunk 2]
```diff
@@ -148,7 +148,7 @@ val ScriptCompilationConfigurationKeys.refineConfigurationOnAnnotations by Prope
 val ScriptCompilationConfigurationKeys.refineConfigurationBeforeCompiling by PropertiesCollection.key<List<RefineConfigurationUnconditionallyData>>(isTransient = true)
 
 /**
- * The list of script fragments that should be compiled intead of the whole text
+ * The list of script fragments that should be compiled instead of the whole text
```
Comment (ligee) [inline libraries/scripting/common/src/kotlin/script/experimental/api/scriptCompilation.kt:151]: Please, to the separate commit (with the "[minor]" prefix), it is completely irrelevant here.

File: plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/repl/ReplCodeAnalyzer.kt [hunk 3]
```diff
@@ -265,11 +265,18 @@ open class ReplCodeAnalyzerBase(
...
             // create scope that wraps previous line lexical scope and adds imports from this line
             val lexicalScopeAfterLastLine = lineInfo.parentLine?.lineDescriptor?.scopeForInitializerResolution ?: return null
             val lastLineImports = lexicalScopeAfterLastLine.parentsWithSelf.first { it is ImportingScope } as ImportingScope
-            val scopesForThisLine = fileScopeFactory.createScopesForFile(lineInfo.linePsi, lastLineImports)
-            val combinedLexicalScopes = lexicalScopeAfterLastLine.replaceImportingScopes(scopesForThisLine.importingScope)
+            val scopesForThisLine = fileScopeFactory.createScopesForFile(linePsi, lastLineImports)
+            val combinedLexicalScopes = if (hasImports)
+                lexicalScopeAfterLastLine.replaceImportingScopes(scopesForThisLine.importingScope)
+            else
+                lexicalScopeAfterLastLine
```
Comment (ligee) [inline plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/repl/ReplCodeAnalyzer.kt:278]: Have you checked that this does not affect implicit imports, that are passed via script compilation configuration?
I believe it shouldn't, but it must be checked, I believe.
You'll need to construct a REPL with a custom script def with implicit imports, to check that. Maybe it makes sense to add tests for it too, we have not many tests with imports and this change can alter the behavior.

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/AbstractLazyMemberScope.kt [hunk 4]
```diff
@@ -188,6 +188,12 @@ protected constructor(
 
     protected abstract fun getNonDeclaredProperties(name: Name, result: MutableSet<PropertyDescriptor>)
 
+    override fun definitelyDoesNotContainName(name: Name): Boolean =
+        !getVariableNames().contains(name) &&
+                !getFunctionNames().contains(name) &&
+                getClassifierNames()?.contains(name) != true &&
```
Comment (dzharkov) [inline compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/AbstractLazyMemberScope.kt:194]: Sorry, it's not obvious but this is incorrect semantic usage of the `getClassifierNames()`: it returns null when the scope can't compute classifier names' set efficiently.
And this happens to `LazyClassMemberScope` unfortunately.
It's necessary to implement it properly there to make `definitelyDoesNotContainName` work correctly

File: plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/repl/ReplCodeAnalyzer.kt (hunk 3 above)
Comment (ileasile) [reply plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/repl/ReplCodeAnalyzer.kt:278]: @ligee are you talking about `defaultImports`, right?

Comment (ileasile) [reply plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/repl/ReplCodeAnalyzer.kt:278]: If so, I've a test

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/AbstractLazyMemberScope.kt (hunk 4 above)
Comment (ileasile) [reply compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/AbstractLazyMemberScope.kt]: @dzharkov Fixed this method, please review new implementation

**Comment by ileasile** [comment]
@ligee @dzharkov I've deleted annotation and related fixes in `TowerResolver`, but added one small thing,
```
override fun definitelyDoesNotContainName(name: Name): Boolean = true
```
for **empty** member scope. It seems to make linear complexity in the test I've added. Maybe it will even suit us in Kotlin kernel, so I ask you to review and merge it.

File: libraries/scripting/common/src/kotlin/script/experimental/api/scriptCompilation.kt (hunk 2 above)
Comment (ileasile) [reply libraries/scripting/common/src/kotlin/script/experimental/api/scriptCompilation.kt:151]: Done

File: core/descriptors/src/org/jetbrains/kotlin/resolve/scopes/MemberScope.kt [hunk 5]
```diff
@@ -46,6 +46,8 @@ interface MemberScope : ResolutionScope {
             p.println("Empty member scope")
         }
 
+        override fun definitelyDoesNotContainName(name: Name): Boolean = true
```
Comment (dzharkov) [inline core/descriptors/src/org/jetbrains/kotlin/resolve/scopes/MemberScope.kt:49]: What is this change for?

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/LazyClassMemberScope.kt [hunk 6]
```diff
@@ -135,6 +135,27 @@ open class LazyClassMemberScope(
...
+    override fun definitelyDoesNotContainName(name: Name): Boolean {
+        if (getVariableNames().contains(name) || getFunctionNames().contains(name))
+            return false
+
+        with(c.syntheticResolveExtension) {
+            if (getSyntheticCompanionObjectNameIfNeeded(thisDescriptor) == name)
+                return false
+            if (getSyntheticFunctionNames(thisDescriptor).contains(name))
+                return false
+            if (getSyntheticNestedClassNames(thisDescriptor).contains(name))
```
Comment (dzharkov) [inline compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/LazyClassMemberScope.kt:147]: Also, I'm concerned we don't handle usual nested classifiers' names here

File: core/descriptors/src/org/jetbrains/kotlin/resolve/scopes/MemberScope.kt (hunk 5 above)
Comment (ileasile) [reply core/descriptors/src/org/jetbrains/kotlin/resolve/scopes/MemberScope.kt:49]: There are empty member scopes created in REPL scripts, so it helps us to filter them out

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/LazyClassMemberScope.kt [hunk 7]
```diff
@@ -135,6 +135,27 @@ open class LazyClassMemberScope(
     override fun getVariableNames() = _variableNames
     override fun getFunctionNames() = _functionNames
 
+    override fun definitelyDoesNotContainName(name: Name): Boolean {
+        if (getVariableNames().contains(name) || getFunctionNames().contains(name))
```
[the remark this answers is not in the document]
Comment (ileasile) [reply compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/LazyClassMemberScope.kt]: Fixed

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/LazyClassMemberScope.kt [hunk 8]
```diff
@@ -135,6 +135,27 @@ open class LazyClassMemberScope(
...
+            return false
+
+        with(c.syntheticResolveExtension) {
+            if (getSyntheticCompanionObjectNameIfNeeded(thisDescriptor) == name)
+                return false
+            if (getSyntheticFunctionNames(thisDescriptor).contains(name))
```
[the remark this answers is not in the document]
Comment (ileasile) [reply compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/LazyClassMemberScope.kt]: Fixed

**Pull request closed without merging**